File: day02/day02.py
#!/usr/bin/env python3

import os
from typing import List

SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))


def read_input(path: str) -> List[str]:
    with open(path, 'r') as f:
        lines = f.read().splitlines()
    return lines



# =========================================

def parse_data(lines):
    result = []
    for line in lines:
        numbers = [int(val) for val in line.split(" ")]
        result.append(numbers)
    return result

def test_stages(stages):
    if stages[1] - stages[0] > 0:
        direction = "+"
    else:
        direction = "-"
    for i in range(len(stages)-1):
        diff = stages[i+1] - stages[i]
        print(diff)
        if diff == 0 or abs(diff) > 3:
            return False
        if diff > 0 and direction == "-":
            return False
        if diff < 0 and direction == "=":
            return False
    return True


def run1(reports):
    count = 0
    for stages in reports:
        print(stages, test_stages(stages))
        if test_stages(stages):
            count += 1
    return count

# =========================================

def run2(lines: List[str]):
    pass


INPUT = f"{SCRIPT_DIR}/input_test.txt"
# INPUT = f"{SCRIPT_DIR}/input.txt"
data = read_input(INPUT)
reports = parse_data(data)

result1 = run1(reports)
print("Result1 = ", result1)

result2 = run2(reports)
print("Result2 = ", result2)

File: day02/day02.rb
#!/usr/bin/env ruby

def parse_data(lines)
  result = []
  lines.each do |line|
    nums = []
    line.split(" ").each {|num| nums.append(num.to_i)}
    result.append(nums)
  end
  return result
end

def is_safe?(stages)
  direction = stages[0] < stages[1] ? "+" : "-"
  for i in 0..(stages.size - 2) do
    return false if stages[i] == stages[i+1]
    diff = stages[i] - stages[i+1]
    return false if diff.abs > 3
    return false if diff > 0 and direction == "+"
    return false if diff < 0 and direction == "-"
  end 
  return true
end

def is_quite_safe?(stages)
  return true if is_safe?(stages)
  for i in 0..(stages.size - 1) do
    test_stages = stages.slice(0,i) + stages.slice(i+1, stages.size)
    return true if is_safe?(test_stages)
  end
  return false
end

def step1(reports)
    count = 0
    reports.each do |stages|
      if is_safe?(stages) then count += 1 end
    end
    return count
end

def step2(reports)
  count = 0
  reports.each do |stages|
    if is_quite_safe?(stages) then count += 1 end
  end
  return count
end


if __FILE__ == $0

    # INPUT_FILE = "input_test.txt"
    INPUT_FILE = "input.txt"
    data = File.read(File.join(__dir__, INPUT_FILE)).split("\n")
    parsed = parse_data(data)

    start = Time.now
    result1 = step1(parsed)
    finish = Time.now
    puts "Result1 = #{result1}   \t(in #{'%.6f' % ((finish - start) * 1000)}ms)"

    start = Time.now
    result2 = step2(parsed)
    finish = Time.now
    puts "Result2 = #{result2}   \t(in #{'%.6f' % ((finish - start) * 1000)}ms)"

end